### PR TITLE
fix(pipeline): Solve data consistency problems of pipeline cron

### PR DESCRIPTION
**Pull request by littlejiancc**
#### What this PR does / why we need it:
Solve data consistency problems of pipeline cron

#### Which issue(s) this PR fixes:

- Fixes #your-issue_number
- [Erda Cloud Issue Link](https://erda.cloud/erda/dop/projects/387/issues/gantt?filter__urlQuery=eyJpdGVyYXRpb25JRHMiOlsxMjgwLDEyNTBdLCJhc3NpZ25lZSI6WyIxMDAxMjYxIl19&id=317624&iterationID=1250&pId=0&type=BUG)

#### Specified Reviewers:

/assign @sfwn 

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English |    Solve data consistency problems of pipeline cron          |
| 🇨🇳 中文    |       解决流水线定时数据一致性问题       |

#### Need cherry-pick to release versions?

Add comment like `/cherry-pick release/1.0` when this PR is merged.

> For details on the cherry pick process, see the [cherry pick requests](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md#how-to-cherry-pick-a-merged-pr) section under [CONTRIBUTING.md](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -49,6 +49,7 @@ require (
 	github.com/elastic/cloud-on-k8s v0.0.0-20210205172912-5ce0eca90c60
 	github.com/elazarl/goproxy v0.0.0-20200421181703-e76ad31c14f6
 	github.com/erda-project/erda-infra v0.0.0-20220607095148-d7ee28ba8e73
+	github.com/erda-project/erda-infra/tools v0.0.0-20220607095148-d7ee28ba8e73 // indirect
```
Comment (sfwn) [inline go.mod:52]: remove it

File: internal/tools/pipeline/providers/cron/cron.service.go [hunk 2]
```diff
@@ -381,6 +384,12 @@ func (s *provider) update(req *pb.CronUpdateRequest, cron db2.PipelineCron, fiel
 			return err
 		}
 	}
+
+	if *cron.Enable {
+		if err := s.Daemon.AddIntoPipelineCrond(&cron); err != nil {
+			return apierrors.ErrUpdatePipelineCron.InternalError(err)
+		}
+	}
```
Comment (sfwn) [inline internal/tools/pipeline/providers/cron/cron.service.go:392]: how about disable

File: internal/tools/pipeline/providers/cron/daemon/provider_test.go [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+
+	"github.com/coreos/etcd/clientv3"
+
+	"github.com/erda-project/erda/apistructs"
+	"github.com/erda-project/erda/bundle"
+	db2 "github.com/erda-project/erda/internal/tools/pipeline/providers/cron/db"
+	"github.com/erda-project/erda/internal/tools/pipeline/providers/edgepipeline_register"
+)
+
+type EdgePipelineRegister struct {
```
Comment (sfwn) [inline internal/tools/pipeline/providers/cron/daemon/provider_test.go:30]: use internal/tools/pipeline/providers/edgepipeline_register/mock.go:25

**Review by sfwn — CHANGES_REQUESTED** [review]

File: internal/tools/pipeline/providers/cron/cron.service.go (hunk 2 above)
Comment (littlejiancc) [reply internal/tools/pipeline/providers/cron/cron.service.go:392]: if cron is diable, no need to update pipeline crond, just update db.

File: internal/tools/pipeline/providers/cron/daemon/provider_test.go (hunk 3 above)
Comment (littlejiancc) [reply internal/tools/pipeline/providers/cron/daemon/provider_test.go:30]: The return value of the mock is different and requires a separate mock

File: go.mod (hunk 1 above)
Comment (littlejiancc) [reply go.mod]: done

File: internal/tools/pipeline/providers/cron/daemon/provider_test.go [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+	"fmt"
+	"net/http"
+	"testing"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/mvcc/mvccpb"
+
+	"github.com/erda-project/erda/apistructs"
+	"github.com/erda-project/erda/bundle"
+	db2 "github.com/erda-project/erda/internal/tools/pipeline/providers/cron/db"
```
Comment (sfwn) [inline internal/tools/pipeline/providers/cron/daemon/provider_test.go:28]: strange db2

File: internal/tools/pipeline/providers/cron/daemon/provider_test.go (hunk 3 above)
Comment (sfwn) [reply internal/tools/pipeline/providers/cron/daemon/provider_test.go:32]: use embed and overwrite some needed methods.

**Review by sfwn — CHANGES_REQUESTED** [review]

File: internal/tools/pipeline/providers/cron/daemon/provider_test.go (hunk 3 above)
Comment (littlejiancc) [reply internal/tools/pipeline/providers/cron/daemon/provider_test.go]: done

**Review by sfwn — APPROVED** [review]

**Pull request merged**